File: skills/ppt-pptx/scripts/smoke_test.py
#!/usr/bin/env python3
"""
smoke_test.py

End-to-end smoke test for the ppt-pptx skill.

Verifies:
- outline -> deck.js -> deck.pptx
- template -> deck.pptx
- sample deck.js -> deck.pptx
- render, overflow, font, and structure checks where applicable
"""

from __future__ import annotations

import argparse
import json
import os
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
SCRIPTS = ROOT / "scripts"
ASSETS = ROOT / "assets"
EXAMPLES = ROOT / "examples"

NODE_DEPS = [
    "pptxgenjs",
    "skia-canvas",
    "linebreak",
    "fontkit",
    "prismjs",
    "mathjax-full",
    "js-yaml",
]

RUST_TOOL_MANIFEST = ROOT.parents[1] / "rust_tools" / "pptx_tool_rs" / "Cargo.toml"


def rust_tool_env() -> dict[str, str]:
    target_root = ROOT.parents[1] / "rust_tools" / "target" / "debug"
    binary = target_root / "pptx_tool_rs"
    env = os.environ.copy()
    env["PPT_PPTX_RUST_TOOL_BIN"] = str(binary)
    env["PPT_PPTX_RUST_TOOL_MANIFEST"] = str(RUST_TOOL_MANIFEST)
    return env


def officecli_available() -> bool:
    return shutil.which("officecli") is not None


def run(
    cmd: list[str],
    cwd: Path,
    label: str,
    *,
    env: dict[str, str] | None = None,
) -> subprocess.CompletedProcess[str]:
    proc = subprocess.run(
        cmd,
        cwd=str(cwd),
        text=True,
        capture_output=True,
        env=env,
    )
    if proc.returncode != 0:
        raise RuntimeError(
            f"{label} failed\n"
            f"cmd: {' '.join(cmd)}\n"
            f"stdout:\n{proc.stdout}\n"
            f"stderr:\n{proc.stderr}"
        )
    return proc


def copy_common_python_tools(dest: Path) -> None:
    for name in [
        "rust_bridge.py",
        "render_slides.py",
        "slides_test.py",
        "detect_font.py",
        "extract_pptx_structure.py",
        "sanitize_pptx.py",
    ]:
        shutil.copy2(SCRIPTS / name, dest / name)


def npm_bootstrap(dest: Path) -> None:
    run(["npm", "init", "-y"], dest, "npm init")
    run(["npm", "install", *NODE_DEPS], dest, "npm install")


def count_pngs(path: Path) -> int:
    return len(list(path.glob("*.png")))


def officecli_doctor(workdir: Path) -> dict | None:
    if not officecli_available():
        return None
    proc = run(
        [sys.executable, str(SCRIPTS / "officecli_bridge.py"), "doctor", "deck.pptx", "--json"],
        workdir,
        "officecli doctor",
    )
    return json.loads(proc.stdout)


def scenario_outline(root: Path) -> dict:
    workdir = root / "outline"
    workdir.mkdir(parents=True, exist_ok=True)

    shutil.copy2(EXAMPLES / "outline_overload.yaml", workdir / "outline.yaml")
    shutil.copy2(SCRIPTS / "outline_to_deck.js", workdir / "outline_to_deck.js")
    shutil.copytree(ASSETS / "pptxgenjs_helpers", workdir / "pptxgenjs_helpers")
    copy_common_python_tools(workdir)
    shutil.copy2(SCRIPTS / "officecli_bridge.py", workdir / "officecli_bridge.py")
    shutil.copy2(SCRIPTS / "hybrid_pipeline.py", workdir / "hybrid_pipeline.py")
    npm_bootstrap(workdir)
    env = rust_tool_env()

    run(["node", "outline_to_deck.js", "outline.yaml", "-o", "deck.js"], workdir, "outline_to_deck")
    run(["node", "deck.js"], workdir, "generated deck.js")
    run(
        [sys.executable, "render_slides.py", "deck.pptx", "--output_dir", "rendered"],
        workdir,
        "render_slides",
        env=env,
    )
    run([sys.executable, "slides_test.py", "deck.pptx"], workdir, "slides_test", env=env)
    run(
        [sys.executable, "detect_font.py", "deck.pptx", "--include-missing", "--include-substituted"],
        workdir,
        "detect_font",
        env=env,
    )
    run(
        [sys.executable, "extract_pptx_structure.py", "deck.pptx", "-o", "structure.json"],
        workdir,
        "extract_structure",
        env=env,
    )
    hybrid = run(
        [sys.executable, "hybrid_pipeline.py", "qa", "deck.pptx", "--rendered-dir", "rendered", "--json"],
        workdir,
        "hybrid qa",
        env=env,
    )
    hybrid_payload = json.loads(hybrid.stdout)

    result = {
        "name": "outline_flow",
        "workdir": str(workdir),
        "deck_exists": (workdir / "deck.pptx").exists(),
        "rendered_pngs": count_pngs(workdir / "rendered"),
        "structure_json": (workdir / "structure.json").exists(),
        "hybrid_render_pngs": hybrid_payload["render"]["png_count"],
    }
    doctor = officecli_doctor(workdir)
    if doctor:
        result["officecli_issue_count"] = doctor["issues"]["count"]
        result["officecli_validation_ok"] = doctor["validation"]["ok"]
    return result


def scenario_template(root: Path) -> dict:
    workdir = root / "template"
    assets_dir = workdir / "assets"
    workdir.mkdir(parents=True, exist_ok=True)
    assets_dir.mkdir(parents=True, exist_ok=True)

    shutil.copy2(ASSETS / "deck.template.js", workdir / "deck.js")
    shutil.copytree(ASSETS / "pptxgenjs_helpers", workdir / "pptxgenjs_helpers")
    copy_common_python_tools(workdir)
    shutil.copy2(SCRIPTS / "officecli_bridge.py", workdir / "officecli_bridge.py")
    shutil.copy2(SCRIPTS / "hybrid_pipeline.py", workdir / "hybrid_pipeline.py")
    npm_bootstrap(workdir)
    env = rust_tool_env()

    run(["node", "deck.js"], workdir, "template deck.js")
    run(
        [sys.executable, "render_slides.py", "deck.pptx", "--output_dir", "rendered"],
        workdir,
        "render_slides",
        env=env,
    )
    run([sys.executable, "slides_test.py", "deck.pptx"], workdir, "slides_test", env=env)
    run(
        [sys.executable, "detect_font.py", "deck.pptx", "--include-missing", "--include-substituted"],
        workdir,
        "detect_font",
        env=env,
    )
    hybrid = run(
        [sys.executable, "hybrid_pipeline.py", "qa", "deck.pptx", "--rendered-dir", "rendered", "--json"],
        workdir,
        "hybrid qa",
        env=env,
    )
    hybrid_payload = json.loads(hybrid.stdout)

    result = {
        "name": "template_flow",
        "workdir": str(workdir),
        "deck_exists": (workdir / "deck.pptx").exists(),
        "rendered_pngs": count_pngs(workdir / "rendered"),
        "hybrid_render_pngs": hybrid_payload["render"]["png_count"],
    }
    doctor = officecli_doctor(workdir)
    if doctor:
        result["officecli_issue_count"] = doctor["issues"]["count"]
        result["officecli_validation_ok"] = doctor["validation"]["ok"]
    return result


def scenario_sample_deck(root: Path) -> dict:
    workdir = root / "sample_deck"
    workdir.mkdir(parents=True, exist_ok=True)

    shutil.copy2(ROOT / "deck.js", workdir / "deck.js")
    shutil.copytree(ASSETS / "pptxgenjs_helpers", workdir / "pptxgenjs_helpers")
    copy_common_python_tools(workdir)
    shutil.copy2(SCRIPTS / "officecli_bridge.py", workdir / "officecli_bridge.py")
    shutil.copy2(SCRIPTS / "hybrid_pipeline.py", workdir / "hybrid_pipeline.py")
    npm_bootstrap(workdir)
    env = rust_tool_env()

    run(["node", "deck.js"], workdir, "sample deck.js")
    run(
        [sys.executable, "render_slides.py", "deck.pptx", "--output_dir", "rendered"],
        workdir,
        "render_slides",
        env=env,
    )
    hybrid = run(
        [sys.executable, "hybrid_pipeline.py", "qa", "deck.pptx", "--rendered-dir", "rendered", "--json"],
        workdir,
        "hybrid qa",
        env=env,
    )
    hybrid_payload = json.loads(hybrid.stdout)

    result = {
        "name": "sample_deck_flow",
        "workdir": str(workdir),
        "deck_exists": (workdir / "deck.pptx").exists(),
        "rendered_pngs": count_pngs(workdir / "rendered"),
        "hybrid_render_pngs": hybrid_payload["render"]["png_count"],
    }
    doctor = officecli_doctor(workdir)
    if doctor:
        result["officecli_issue_count"] = doctor["issues"]["count"]
        result["officecli_validation_ok"] = doctor["validation"]["ok"]
    return result


def main() -> int:
    parser = argparse.ArgumentParser(description="Run end-to-end smoke tests for skills/ppt-pptx.")
    parser.add_argument("--keep-workdir", action="store_true", help="Keep the temporary workspace.")
    parser.add_argument("--json", action="store_true", help="Emit machine-readable JSON summary.")
    parser.add_argument(
        "--strict-officecli",
        action="store_true",
        help="Fail if optional OfficeCLI doctor finds issues or validation failures.",
    )
    args = parser.parse_args()

    temp_dir_obj = tempfile.TemporaryDirectory(prefix="pptx-smoke-")
    temp_root = Path(temp_dir_obj.name)

    try:
        run(
            ["cargo", "build", "--manifest-path", str(RUST_TOOL_MANIFEST)],
            ROOT.parents[1],
            "cargo build pptx_tool_rs",
        )
        results = [
            scenario_outline(temp_root),
            scenario_template(temp_root),
            scenario_sample_deck(temp_root),
        ]
        payload = {
            "status": "pass",
            "root": str(temp_root),
            "officecli_available": officecli_available(),
            "results": results,
        }
        if args.strict_officecli and any(
            item.get("officecli_issue_count", 0) > 0 or item.get("officecli_validation_ok") is False
            for item in results
        ):
            raise RuntimeError("OfficeCLI strict audit found deck issues or validation failures")
        if args.json:
            print(json.dumps(payload, ensure_ascii=False, indent=2))
        else:
            print(f"PASS: ppt-pptx smoke test ({temp_root})")
            for item in results:
                extras = ", ".join(f"{k}={v}" for k, v in item.items() if k not in {"name", "workdir"})
                print(f"- {item['name']}: {extras}")

        if args.keep_workdir:
            print(f"Kept workspace: {temp_root}")
        return 0
    except Exception as exc:
        if args.json:
            print(json.dumps({"status": "fail", "root": str(temp_root), "error": str(exc)}, ensure_ascii=False, indent=2))
        else:
            print(f"FAIL: {exc}", file=sys.stderr)
            print(f"Workspace: {temp_root}", file=sys.stderr)
        if args.keep_workdir:
            print(f"Kept workspace: {temp_root}", file=sys.stderr)
        return 1
    finally:
        if not args.keep_workdir:
            temp_dir_obj.cleanup()


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/ppt-pptx/scripts/smoke_test.js
#!/usr/bin/env node

const fs = require("node:fs");
const os = require("node:os");
const path = require("node:path");
const { spawnSync } = require("node:child_process");

const ROOT = path.resolve(__dirname, "..");
const SCRIPTS = path.join(ROOT, "scripts");
const ASSETS = path.join(ROOT, "assets");
const EXAMPLES = path.join(ROOT, "examples");
const TOOL_RUNNER = path.join(SCRIPTS, "pptx_tool.js");
const RUST_TOOL_BIN = path.resolve(ROOT, "..", "..", "rust_tools", "target", "debug", "pptx_tool_rs");
const RUST_TOOL_MANIFEST = path.resolve(ROOT, "..", "..", "rust_tools", "pptx_tool_rs", "Cargo.toml");

const NODE_DEPS = [
  "pptxgenjs",
  "skia-canvas",
  "linebreak",
  "fontkit",
  "prismjs",
  "mathjax-full",
  "js-yaml",
];

function rustToolEnv() {
  return {
    ...process.env,
    PPT_PPTX_RUST_TOOL_BIN: RUST_TOOL_BIN,
    PPT_PPTX_RUST_TOOL_MANIFEST: RUST_TOOL_MANIFEST,
  };
}

function rustToolCmd(...args) {
  return ["node", TOOL_RUNNER, ...args];
}

function run(cmd, cwd, label, env = process.env) {
  const proc = spawnSync(cmd[0], cmd.slice(1), {
    cwd,
    env,
    encoding: "utf8",
  });
  if (proc.status !== 0) {
    throw new Error(
      `${label} failed\ncmd: ${cmd.join(" ")}\nstdout:\n${proc.stdout || ""}\nstderr:\n${proc.stderr || ""}`,
    );
  }
  return proc;
}

function parseJsonOutput(label, text) {
  const trimmed = text.trim();
  const starts = [];
  for (let index = 0; index < trimmed.length; index += 1) {
    const ch = trimmed[index];
    if (ch === "{" || ch === "[") {
      starts.push(index);
    }
  }
  for (const start of starts) {
    try {
      return JSON.parse(trimmed.slice(start));
    } catch {}
  }
  throw new Error(`${label} did not return valid JSON\nstdout:\n${text}`);
}

function copyFile(src, dest) {
  fs.mkdirSync(path.dirname(dest), { recursive: true });
  fs.copyFileSync(src, dest);
}

function copyToolRunner(dest) {
  copyFile(TOOL_RUNNER, path.join(dest, "scripts", "pptx_tool.js"));
}

function copyDir(src, dest) {
  fs.cpSync(src, dest, { recursive: true });
}

function npmBootstrap(dest) {
  run(["npm", "init", "-y"], dest, "npm init");
  run(["npm", "install", ...NODE_DEPS], dest, "npm install");
}

function countPngs(dir) {
  if (!fs.existsSync(dir)) {
    return 0;
  }
  return fs.readdirSync(dir).filter((name) => name.endsWith(".png")).length;
}

function officecliAvailable() {
  const proc = spawnSync("which", ["officecli"], { encoding: "utf8" });
  return proc.status === 0;
}

function officecliDoctor(workdir) {
  if (!officecliAvailable()) {
    return null;
  }
  const proc = run(
    rustToolCmd("office", "doctor", "deck.pptx", "--json"),
    workdir,
    "officecli doctor",
    rustToolEnv(),
  );
  return parseJsonOutput("officecli doctor", proc.stdout);
}

function enrichWithDoctor(result, workdir) {
  const doctor = officecliDoctor(workdir);
  if (doctor) {
    result.officecli_issue_count = doctor.issues.count;
    result.officecli_validation_ok = doctor.validation.ok;
  }
  return result;
}

function scenarioOutline(root) {
  const workdir = path.join(root, "outline");
  fs.mkdirSync(workdir, { recursive: true });

  copyFile(path.join(EXAMPLES, "outline_overload.yaml"), path.join(workdir, "outline.yaml"));
  copyFile(path.join(SCRIPTS, "outline_to_deck.js"), path.join(workdir, "outline_to_deck.js"));
  copyDir(path.join(ASSETS, "pptxgenjs_helpers"), path.join(workdir, "pptxgenjs_helpers"));
  copyToolRunner(workdir);
  npmBootstrap(workdir);

  const env = rustToolEnv();
  run(["node", "outline_to_deck.js", "outline.yaml", "-o", "deck.js"], workdir, "outline_to_deck", env);
  run(["node", "deck.js"], workdir, "generated deck.js", env);
  run(rustToolCmd("render", "deck.pptx", "--output_dir", "rendered"), workdir, "render", env);
  run(rustToolCmd("slides-test", "deck.pptx"), workdir, "slides_test", env);
  run(
    rustToolCmd("detect-fonts", "deck.pptx", "--include-missing", "--include-substituted"),
    workdir,
    "detect_font",
    env,
  );
  run(rustToolCmd("extract-structure", "deck.pptx", "-o", "structure.json"), workdir, "extract_structure", env);
  const hybrid = run(
    rustToolCmd("qa", "deck.pptx", "--rendered-dir", "rendered", "--json"),
    workdir,
    "hybrid qa",
    env,
  );
  const hybridPayload = parseJsonOutput("hybrid qa", hybrid.stdout);

  return enrichWithDoctor(
    {
      name: "outline_flow",
      workdir,
      deck_exists: fs.existsSync(path.join(workdir, "deck.pptx")),
      rendered_pngs: countPngs(path.join(workdir, "rendered")),
      structure_json: fs.existsSync(path.join(workdir, "structure.json")),
      hybrid_render_pngs: hybridPayload.render.png_count,
    },
    workdir,
  );
}

function scenarioTemplate(root) {
  const workdir = path.join(root, "template");
  fs.mkdirSync(path.join(workdir, "assets"), { recursive: true });

  copyFile(path.join(ASSETS, "deck.template.js"), path.join(workdir, "deck.js"));
  copyDir(path.join(ASSETS, "pptxgenjs_helpers"), path.join(workdir, "pptxgenjs_helpers"));
  copyToolRunner(workdir);
  npmBootstrap(workdir);

  const env = rustToolEnv();
  run(["node", "deck.js"], workdir, "template deck.js", env);
  run(rustToolCmd("render", "deck.pptx", "--output_dir", "rendered"), workdir, "render", env);
  run(rustToolCmd("slides-test", "deck.pptx"), workdir, "slides_test", env);
  run(
    rustToolCmd("detect-fonts", "deck.pptx", "--include-missing", "--include-substituted"),
    workdir,
    "detect_font",
    env,
  );
  const hybrid = run(
    rustToolCmd("qa", "deck.pptx", "--rendered-dir", "rendered", "--json"),
    workdir,
    "hybrid qa",
    env,
  );
  const hybridPayload = parseJsonOutput("hybrid qa", hybrid.stdout);

  return enrichWithDoctor(
    {
      name: "template_flow",
      workdir,
      deck_exists: fs.existsSync(path.join(workdir, "deck.pptx")),
      rendered_pngs: countPngs(path.join(workdir, "rendered")),
      hybrid_render_pngs: hybridPayload.render.png_count,
    },
    workdir,
  );
}

function scenarioSampleDeck(root) {
  const workdir = path.join(root, "sample_deck");
  fs.mkdirSync(workdir, { recursive: true });

  copyFile(path.join(ROOT, "deck.js"), path.join(workdir, "deck.js"));
  copyDir(path.join(ASSETS, "pptxgenjs_helpers"), path.join(workdir, "pptxgenjs_helpers"));
  copyToolRunner(workdir);
  npmBootstrap(workdir);

  const env = rustToolEnv();
  run(["node", "deck.js"], workdir, "sample deck.js", env);
  run(rustToolCmd("render", "deck.pptx", "--output_dir", "rendered"), workdir, "render", env);
  const hybrid = run(
    rustToolCmd("qa", "deck.pptx", "--rendered-dir", "rendered", "--json"),
    workdir,
    "hybrid qa",
    env,
  );
  const hybridPayload = parseJsonOutput("hybrid qa", hybrid.stdout);

  return enrichWithDoctor(
    {
      name: "sample_deck_flow",
      workdir,
      deck_exists: fs.existsSync(path.join(workdir, "deck.pptx")),
      rendered_pngs: countPngs(path.join(workdir, "rendered")),
      hybrid_render_pngs: hybridPayload.render.png_count,
    },
    workdir,
  );
}

function parseArgs(argv) {
  return {
    keepWorkdir: argv.includes("--keep-workdir"),
    json: argv.includes("--json"),
    strictOfficecli: argv.includes("--strict-officecli"),
  };
}

function printTextSummary(root, results) {
  console.log(`PASS: ppt-pptx smoke test (${root})`);
  for (const item of results) {
    const extras = Object.entries(item)
      .filter(([key]) => key !== "name" && key !== "workdir")
      .map(([key, value]) => `${key}=${value}`)
      .join(", ");
    console.log(`- ${item.name}: ${extras}`);
  }
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const tempRoot = fs.mkdtempSync(path.join(os.tmpdir(), "pptx-smoke-"));

  try {
    if (!fs.existsSync(RUST_TOOL_BIN)) {
      run(["cargo", "build", "--manifest-path", RUST_TOOL_MANIFEST], path.resolve(ROOT, "..", ".."), "cargo build");
    }
    const results = [
      scenarioOutline(tempRoot),
      scenarioTemplate(tempRoot),
      scenarioSampleDeck(tempRoot),
    ];
    if (
      args.strictOfficecli &&
      results.some(
        (item) => (item.officecli_issue_count || 0) > 0 || item.officecli_validation_ok === false,
      )
    ) {
      throw new Error("OfficeCLI strict audit found deck issues or validation failures");
    }

    const payload = {
      status: "pass",
      root: tempRoot,
      officecli_available: officecliAvailable(),
      results,
    };

    if (args.json) {
      console.log(JSON.stringify(payload, null, 2));
    } else {
      printTextSummary(tempRoot, results);
      if (args.keepWorkdir) {
        console.log(`Kept workspace: ${tempRoot}`);
      }
    }
    return 0;
  } catch (error) {
    if (args.json) {
      console.log(
        JSON.stringify(
          {
            status: "fail",
            root: tempRoot,
            error: error instanceof Error ? error.message : String(error),
          },
          null,
          2,
        ),
      );
    } else {
      console.error(`FAIL: ${error instanceof Error ? error.message : String(error)}`);
      console.error(`Workspace: ${tempRoot}`);
      if (args.keepWorkdir) {
        console.error(`Kept workspace: ${tempRoot}`);
      }
    }
    return 1;
  } finally {
    if (!args.keepWorkdir) {
      fs.rmSync(tempRoot, { recursive: true, force: true });
    }
  }
}

process.exit(main());
